sw/device/tests/sim_dv/uart_tx_rx_test.c

```diff
@@ -0,0 +1,350 @@
+  CHECK(status, "dif_plic_irq_trigger_type_set failed");
+
+  // Set the priority of UART interrupts at PLIC to be >=1 (so ensure the target
+  // does get interrupted).
+  status = dif_plic_irq_priority_set(&plic0,
+                                     kTopEarlgreyPlicIrqIdUartTxWatermark, 0x1);
+  status &= dif_plic_irq_priority_set(
+      &plic0, kTopEarlgreyPlicIrqIdUartRxWatermark, 0x2);
+  status &=
+      dif_plic_irq_priority_set(&plic0, kTopEarlgreyPlicIrqIdUartTxEmpty, 0x3);
+  status &= dif_plic_irq_priority_set(&plic0,
+                                      kTopEarlgreyPlicIrqIdUartRxOverflow, 0x1);
+  status &= dif_plic_irq_priority_set(&plic0,
+                                      kTopEarlgreyPlicIrqIdUartRxFrameErr, 0x2);
+  status &= dif_plic_irq_priority_set(&plic0,
+                                      kTopEarlgreyPlicIrqIdUartRxBreakErr, 0x3);
+  status &= dif_plic_irq_priority_set(&plic0,
+                                      kTopEarlgreyPlicIrqIdUartRxTimeout, 0x1);
+  status &= dif_plic_irq_priority_set(
+      &plic0, kTopEarlgreyPlicIrqIdUartRxParityErr, 0x2);
+  CHECK(status, "dif_plic_irq_priority_set failed");
+
+  // Set the threshold for the Ibex to 0.
+  status =
+      dif_plic_target_threshold_set(&plic0, kTopEarlgreyPlicTargetIbex0, 0x0);
+  CHECK(status, "dif_plic_target_threshold_set failed");
+
+  // Enable all UART interrupts at the PLIC.
+  status = dif_plic_irq_enable_set(&plic0, kTopEarlgreyPlicIrqIdUartTxWatermark,
+                                   kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
+  status &=
+      dif_plic_irq_enable_set(&plic0, kTopEarlgreyPlicIrqIdUartRxWatermark,
+                              kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
+  status &=
+      dif_plic_irq_enable_set(&plic0, kTopEarlgreyPlicIrqIdUartTxEmpty,
+                              kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
+  status &=
+      dif_plic_irq_enable_set(&plic0, kTopEarlgreyPlicIrqIdUartRxOverflow,
+                              kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
+  status &=
+      dif_plic_irq_enable_set(&plic0, kTopEarlgreyPlicIrqIdUartRxFrameErr,
+                              kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
+  status &=
+      dif_plic_irq_enable_set(&plic0, kTopEarlgreyPlicIrqIdUartRxBreakErr,
+                              kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
+  status &=
+      dif_plic_irq_enable_set(&plic0, kTopEarlgreyPlicIrqIdUartRxTimeout,
+                              kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
+  status &=
+      dif_plic_irq_enable_set(&plic0, kTopEarlgreyPlicIrqIdUartRxParityErr,
+                              kTopEarlgreyPlicTargetIbex0, kDifPlicEnable);
+  CHECK(status, "dif_plic_irq_enable_set failed");
+
+  // Enable IRQs at Ibex.
+  irq_global_ctrl(true);
+  irq_external_ctrl(true);
+
+  // Keep sending and receiving data until we are done.
+  bool uart_tx_done = false;
```

Comment:
Wouldn't it make more sense to just compare `uart_tx_bytes_written` and `uart_rx_bytes_written` to `UART_DATASET_SIZE` in the `while` loop and in the two `if` statements below? That way you can get rid of the comparison at the end of each if clause.